slides/advanced-01-feature-engineering.qmd

```diff
@@ -0,0 +1,783 @@
+
+agent_stats %>%   
+  ggplot(aes(x = ADR)) +
+  geom_histogram(bins = 30, col = "red", fill = "red", alpha = 1/3) +
+  labs(x = "Average ADR per agent")
+```
+:::
+:::
+
+## Collapsing factor levels
+
+There is a recipe step that will redefine factor levels based on the their frequency in the training set: 
+
+```{r}
+#| label: step-other
+#| code-line-numbers: "3:4"
+
+hotel_other_rec <-
+  recipe(avg_price_per_room ~ ., data = hotel_tr) %>% 
+  step_YeoJohnson(lead_time) %>%
+  step_other(agent, threshold = 0.001) %>%
+  step_dummy(all_nominal_predictors()) %>%
+  step_zv(all_predictors())
+```
+
+```{r}
+#| label: other-res 
+#| include: false
+
+retained_agents <-
+  recipe(avg_price_per_room ~ ., data = hotel_tr) %>%
+  step_mutate(original = agent) %>% 
+  step_other(agent, threshold = 0.001) %>%
+  step_dummy(all_nominal_predictors()) %>%
+  step_zv(all_predictors()) %>% 
+  prep() %>% 
+  tidy(number = 2)
+
+num_agents <- length(unique(hotel_tr$agent))
+num_other <- num_agents - length(retained_agents$retained)
+```
+
+Using this code, `r num_other` agents (out of `r num_agents`) were collapsed into "other" based on the training set.
+
+We _could_ try to optimize the threshold for collapsing (see the next set of slides on model tuning).
+
+## Does othering help? 
+
+```{r}
+#| label: update-recipe
+#| cache: true
+#| code-line-numbers: "3|"
+hotel_other_wflow <-
+  hotel_lm_wflow %>%
+  update_recipe(hotel_other_rec)
+
+hotel_other_res <-
+  hotel_other_wflow %>%
+  fit_resamples(hotel_rs, control = ctrl)
+
+collect_metrics(hotel_other_res)
+```
+
+A little better MAE and much faster to complete.  
+
+Now let's look at a more sophisticated tool called effect feature hashing. 
+
+## Feature Hashing
+
+Between `agent` and `company`, simple dummy variables would create `r length(unique(hotel_tr$agent)) + length(unique(hotel_tr$company))` new columns (that are mostly zeros).
+
+Another option is to have a binary indicator that combines some levels of these variables.
+
+Feature hashing ([_FES_](https://bookdown.org/max/FES/encoding-predictors-with-many-categories.html),[_SMLTAR_](https://smltar.com/mlregression.html#case-study-feature-hashing), [_TMwR_](https://www.tmwr.org/categorical.html#feature-hashing)):  
+
+- uses the character value of the levels 
+- converts them to integer hash value
+- uses the integer to assign them to a specific indicator column. 
+
+## Feature Hashing
+
+Suppose we want to use 32 indicator variables for `agent`. 
+
+For a agent with value "`Max_Kuhn`", a hashing function converts it to an integer (say `r strtoi(substr(rlang::hash("Max_Kuhn"), 26, 32), 16)`). 
+
+To assign it to one of the 32 columns, we would use modular arithmetic to assign it to a column: 
+
+```{r}
+#| label: hash
+
+# For "Max_Kuhn" but a '1' in column: 
+210397726 %% 32
+```
+
+[Hash functions](https://www.metamorphosite.com/one-way-hash-encryption-sha1-data-software) are used in cryptology and are meant to _emulate_ randomness. 
```

Comment:
while this is true, the hashing function used in feature hashing are not cryptologically safe, rather they are fast and uniform.

Reply:
Good point.